File: 2334-number-of-flowers-in-full-bloom/number-of-flowers-in-full-bloom.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
        people = [(p , i) for i , p in enumerate(people)]
        res = [0] * len(people)
        flowers.sort()
        end =[]

        j = 0

        for p , i in sorted(people):
            while j < len(flowers) and flowers[j][0] <= p:
                heapq.heappush(end ,flowers[j][1])
                j += 1

            while end and end[0] < p:
                heapq.heappop(end)

            res[i] = len(end)

        return res
        

        
        # people =[(p,i) for i,p in enumerate(people)]
        # res = [0] * len(people)

        # start = [f[0] for f in flowers]
        # end = [f[1] for f in flowers]
        # heapq.heapify(start)
        # heapq.heapify(end)
        # count = 0  

        # for p , i in sorted(people):

        #     while start and start[0] <= p:
        #         heapq.heappop(start)
        #         count += 1
        #     while end and end[0] < p:
        #         heapq.heappop(end)
        #         count -= 1
        #     res[i] = count

        # return res

        

File: 2334-number-of-flowers-in-full-bloom/number-of-flowers-in-full-bloom.java
class Solution {
    public int[] fullBloomFlowers(int[][] flowers, int[] people) {

        int n = people.length;
        int[] res = new int[n];

        int[][] persons = new int[n][2];
        for (int i = 0 ; i < n ; i++)
        {
            persons[i][0] = people[i];
            persons[i][1] = i ;
        }
        Arrays.sort(persons , (a,b) -> Integer.compare(a[0] , b[0])) ;
        Arrays.sort(flowers , (a,b) -> Integer.compare(a[0] , b[0])) ;

        PriorityQueue<Integer> heap = new PriorityQueue<>();

        int j = 0 ;

        for(int[] person : persons)
        {
            int p = person[0] ;
            int idx = person[1] ;

            while( j < flowers.length && flowers[j][0] <= p)
            {
                heap.offer(flowers[j][1]) ;
                j ++ ;
            }
            while( !heap.isEmpty() && heap.peek() < p)
            {
                heap.poll() ;
            }
            res[idx] = heap.size();
        }

        return res;

        

    }
}